include/util/cell_storage.hpp

```diff
@@ -0,0 +1,295 @@
+    using BoundaryOffset = std::uint32_t;
+    using BoundarySize = std::uint32_t;
+    using SourceIndex = std::uint32_t;
+    using DestinationIndex = std::uint32_t;
+
+    static constexpr auto INVALID_WEIGHT_OFFSET = std::numeric_limits<WeightOffset>::max();
+    static constexpr auto INVALID_BOUNDARY_OFFSET = std::numeric_limits<BoundaryOffset>::max();
+
+  private:
+    struct CellData
+    {
+        WeightOffset weight_offset = INVALID_WEIGHT_OFFSET;
+        BoundaryOffset source_boundary_offset = INVALID_BOUNDARY_OFFSET;
+        BoundaryOffset destination_boundary_offset = INVALID_BOUNDARY_OFFSET;
+        BoundarySize num_source_nodes = 0;
+        BoundarySize num_destination_nodes = 0;
+    };
+
+    // Implementation of the cell view. We need a template parameter here
+    // because we need to derive a read-only and read-write view from this.
+    template <typename WeightPtrT> class CellImpl
+    {
+      private:
+        using WeightValueT = EdgeWeight;
+        using WeightRefT = decltype(*WeightPtrT());
+        BoundarySize num_source_nodes;
+        BoundarySize num_destination_nodes;
+
+        WeightPtrT const weights;
+        const NodeID *const source_boundary;
+        const NodeID *const destination_boundary;
+
+        using RowIterator = WeightPtrT;
+        class ColumnIterator : public std::iterator<std::random_access_iterator_tag, EdgeWeight>
+        {
+          public:
+            explicit ColumnIterator(WeightPtrT begin, std::size_t row_length)
+                : current(begin), offset(row_length)
+            {
+            }
+
+            WeightRefT operator*() const { return *current; }
+
+            ColumnIterator &operator++()
+            {
+                current += offset;
+                return *this;
+            }
+
+            ColumnIterator &operator+=(int amount)
+            {
+                current += offset * amount;
+                return *this;
+            }
+
+            bool operator==(const ColumnIterator& other) const
+            {
+                return current == other.current;
+            }
+
+            bool operator!=(const ColumnIterator& other) const
+            {
+                return current != other.current;
+            }
+
+            int operator-(const ColumnIterator &other) const { return (current - other.current) / offset; }
+
+          private:
+            WeightPtrT current;
+            std::size_t offset;
+        };
+
+        std::size_t GetRow(NodeID node) const
+        {
+            return std::find(source_boundary, source_boundary + num_source_nodes, node) -
+                   source_boundary;
+        }
+        std::size_t GetColumn(NodeID node) const
+        {
+            return std::find(
+                       destination_boundary, destination_boundary + num_destination_nodes, node) -
+                   destination_boundary;
+        }
+
+      public:
+        std::pair<RowIterator, RowIterator> GetOutWeight(NodeID node) const
```

Comment:
Here and below: for all the pairs of iterators, what about either using

    boost:make_iterator_range(first, last)

or creating a small struct with `begin` and `end` functions so we can iterate over the return values and use std algorithms directly.